sentry-core/src/main/java/io/sentry/transport/HttpTransport.java

```diff
@@ -0,0 +1,186 @@
+import java.util.function.Consumer;
+import javax.net.ssl.HttpsURLConnection;
+
+/**
+ * An implementation of the {@link ITransport} interface that sends the events to the Sentry server
+ * over HTTP(S) in UTF-8 encoding.
+ */
+public class HttpTransport implements ITransport {
+  public static final int HTTP_TOO_MANY_REQUESTS = 429;
+
+  @Nullable private final Proxy proxy;
+  private final Consumer<URLConnection> requestUpdater;
+  private final int connectionTimeout;
+  private final int readTimeout;
+  private final boolean bypassSecurity;
+  private final URL sentryUrl;
+  private final SentryOptions options;
+
+  /**
+   * Constructs a new HTTP transport instance. Notably, the provided {@code requestUpdater} must set
+   * the appropriate content encoding header for the {@link io.sentry.ISerializer} instance obtained
+   * from the options.
+   *
+   * @param options sentry options to read the config from
+   * @param proxy the proxy to use, if any
+   * @param requestUpdater this consumer is given a chance to set up the request before it is sent
+   * @param connectionTimeout connection timeout
+   * @param readTimeout read timeout
+   * @param bypassSecurity whether to ignore TLS errors
+   * @throws URISyntaxException when options contain invalid DSN
+   * @throws MalformedURLException when options contain invalid DSN
+   */
+  public HttpTransport(
+      SentryOptions options,
+      @Nullable Proxy proxy,
+      Consumer<URLConnection> requestUpdater,
+      int connectionTimeout,
+      int readTimeout,
+      boolean bypassSecurity)
+      throws URISyntaxException, MalformedURLException {
+    this.proxy = proxy;
+    this.requestUpdater = requestUpdater;
+    this.connectionTimeout = connectionTimeout;
+    this.readTimeout = readTimeout;
+    this.options = options;
+    this.sentryUrl = new URI(options.getDsn()).toURL();
+    this.bypassSecurity = bypassSecurity;
+  }
+
+  // visible for testing
+  // giving up on testing this method is probably the simplest way of having the rest of the class
+  // testable...
+  protected HttpURLConnection open(URL url, Proxy proxy) throws IOException {
+    return (HttpURLConnection)
+        (proxy == null ? sentryUrl.openConnection() : sentryUrl.openConnection(proxy));
+  }
+
+  @Override
+  public TransportResult send(SentryEvent event, ISerializer serializer) throws IOException {
+    HttpURLConnection connection = open(sentryUrl, proxy);
+    requestUpdater.accept(connection);
+
+    connection.setRequestMethod("POST");
+    connection.setDoOutput(true);
+    connection.setRequestProperty("Content-Encoding", "UTF-8");
+    connection.setConnectTimeout(connectionTimeout);
+    connection.setReadTimeout(readTimeout);
+
+    if (bypassSecurity && connection instanceof HttpsURLConnection) {
+      ((HttpsURLConnection) connection).setHostnameVerifier((__, ___) -> true);
+    }
+
+    connection.connect();
+
+    try (OutputStream outputStream = connection.getOutputStream()) {
```

Comment:
*try-with-resources* is supported on all Android API levels ONLY when using AGP >= 3.0.0

Reply:
we are using AGP 3.5.0

Reply:
This is true. But as far as I know the Java library (or an aar) will include Java bytecode classes which will be dexed when the consumer APK is built. What if the consumer APK is built with a version of AGP that does not use D8?
I might be completely wrong here, but it should be pretty easy to check.

Reply:
Small test from a Java class in an android library aar built with AGP 3.5.0.
Added a try-with-resources statement and checked the java bytecode in the relative class file:
```
     144: invokevirtual #124                // Method it/immobiliare/android/Test.close:()V
     147: goto          211
     150: astore        4
     152: aload_3
     153: aload         4
     155: invokevirtual #126                // Method java/lang/Throwable.addSuppressed:(Ljava/lang/Throwable;)V
     158: goto          211
     161: aload_2
     162: invokevirtual #124                // Method it/immobiliare/android/Test.close:()V
```

As you can see javac adds this line:
`     155: invokevirtual #126                // Method java/lang/Throwable.addSuppressed:(Ljava/lang/Throwable;)V`
which needs D8 desugaring to work on API < 19.

Reply:
it's worth to be checked, but I believe as soon as the lib is built, it doesn't matter who consumes it, our AGP/D8/R8 already desugared it, but if somebody else adds it as a dependency and don't use D8/R8/AGP >= 3, might be a problem, otherwise all Java 8 features will have the same problem.

Reply:
@ninniuz thanks for checking it out, I also checked at https://android.googlesource.com/platform/external/desugar/+/master/java/com/google/devtools/build/android/desugar/runtime/ThrowableExtension.java
After `aar` is assembled using AGP 3 >= 0 with D8/R8, but correct me if I am wrong, does the client still needs to have desugaring enabled? don't think so, right?

Reply:
The bytecode I was showing you was extracted from a `.class` file inside an `aar`, so I can state that desugaring is not applied.

I think your statement `otherwise all Java 8 features will have the same problem.` is correct.
It might be worth to check what happens on a test apk which uses AGP < 3.0.0

Reply:
Huh, for reasons unbeknownst to me, I had Android support disabled in Intellij. I now see the squiggly lines! :)